File: src/decoders/evaluation.ts
import type { Evaluation, Skill } from "~/models";
import { decodePronoteDate } from "./pronote-date";
import { decodeSubject } from "./subject";
import { decodeSkill } from "./skill";

export const decodeEvaluation = (evaluation: any): Evaluation => {
  const skills: Skill[] = evaluation.listeNiveauxDAcquisitions.V.map(decodeSkill);
  skills.sort((skillA, skillB) => skillA.order - skillB.order);

  return {
    skills,
    name: evaluation.L,
    id: evaluation.N,
    teacher: evaluation.individu.V.L,
    coefficient: evaluation.coefficient,
    description: evaluation.descriptif,
    subject: decodeSubject(evaluation.matiere.V),
    levels: evaluation.listePaliers.V.map((level: any) => level.L),
    date: decodePronoteDate(evaluation.date.V)
  };
};

File: library/src/commonMain/kotlin/decoders/evaluation.kt
package decoders

import kotlinx.serialization.json.*
import models.Evaluation

fun decodeEvaluation (evaluation: JsonObject): Evaluation {
    val skills = evaluation["listeNiveauxDAcquisitions"]!!.jsonObject["V"]!!.jsonArray.map { decodeSkill(it.jsonObject) }
    skills.sortedWith { a, b -> a.order - b.order }

    return Evaluation(
        skills = skills,
        name = evaluation["L"]!!.jsonPrimitive.content,
        id = evaluation["N"]!!.jsonPrimitive.content,
        teacher = evaluation["individu"]!!.jsonObject["V"]!!.jsonObject["L"]!!.jsonPrimitive.content,
        coefficient = evaluation["coefficient"]!!.jsonPrimitive.int,
        description = evaluation["descriptif"]!!.jsonPrimitive.content,
        subject = decodeSubject(evaluation["matiere"]!!.jsonObject["V"]!!.jsonObject),
        levels = evaluation["listePaliers"]!!.jsonObject["V"]!!.jsonArray.map { it.jsonObject["L"]!!.jsonPrimitive.content },
        date = decodePronoteDate(evaluation["date"]!!.jsonObject["V"]!!.jsonPrimitive.content)
    )
}